account_invoice_restrict_linked_so/views/account_move.xml

```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+
+    <record id="view_move_form_inherit" model="ir.ui.view">
+        <field name="name">account.move.form.inherit</field>
+        <field name="model">account.move</field>
+        <field name="inherit_id" ref="account.view_move_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='invoice_line_ids']/tree" position="inside">
+                <field name="from_so_line" invisible="1" />
+            </xpath>
+            <xpath
+                expr="//field[@name='invoice_line_ids']/tree/field[@name='product_id']"
+                position="attributes"
+            >
+                <attribute
+                    name="attrs"
+                >{'readonly':[('from_so_line','!=',False)]}</attribute>
```

Comment:
The `from_so_line` field bring little benefit.
You can just check if `sale_line_ids` is set or empty.

```suggestion
                >{'readonly':[('sale_line_ids','!=',False)]}</attribute>
```